be/src/olap/rowset/beta_rowset_writer.cpp

```diff
@@ -102,6 +110,284 @@ Status BetaRowsetWriter::add_block(const vectorized::Block* block) {
+    }
+    return Status::OK();
+}
+
+void BetaRowsetWriter::find_longest_consecutive_small_segment(
+        SegCompactionCandidatesSharedPtr segments) {
+    std::vector<segment_v2::SegmentSharedPtr> all_segments;
+    load_noncompacted_segments(&all_segments);
+
+    std::stringstream ss_all;
+    for (auto& segment : all_segments) {
+        ss_all << "[id:" << segment->id() << " num_rows:" << segment->num_rows() << "]";
+    }
+    LOG(INFO) << "all noncompacted segments num:" << all_segments.size()
+              << " list of segments:" << ss_all.str();
+
+    bool is_terminated_by_big = false;
+    bool let_big_terminate = false;
+    size_t small_threshold = config::segcompaction_small_threshold;
+    for (int64_t i = 0; i < all_segments.size(); ++i) {
+        segment_v2::SegmentSharedPtr seg = all_segments[i];
+        if (seg->num_rows() > small_threshold) {
+            if (let_big_terminate) {
+                is_terminated_by_big = true;
+                break;
+            } else {
+                rename_compacted_segment_plain(_segcompacted_point);
+                ++_segcompacted_point;
+            }
+        } else {
+            let_big_terminate = true; // break if find a big after small
+            segments->push_back(seg);
+            ++_segcompacted_point;
+        }
+    }
+    size_t s = segments->size();
+    if (!is_terminated_by_big && s <= (config::segcompaction_threshold_segment_num / 2)) {
+        // start with big segments and end with small, better to do it in next
+        // round to compact more at once
+        _segcompacted_point -= s;
+        segments->clear();
+        LOG(INFO) << "candidate segments num too small:" << s;
+        return;
+    }
+    if (s == 1) { // poor bachelor, let it go
+        LOG(INFO) << "only one candidate segment";
+        rename_compacted_segment_plain(_segcompacted_point - 1);
+        segments->clear();
+        return;
+    }
+    std::stringstream ss;
+    for (auto& segment : (*segments.get())) {
+        ss << "[id:" << segment->id() << " num_rows:" << segment->num_rows() << "]";
+    }
+    LOG(INFO) << "candidate segments num:" << s << " list of candidates:" << ss.str();
+}
+
+SegCompactionCandidatesSharedPtr BetaRowsetWriter::get_segcompaction_candidates(bool is_last) {
+    SegCompactionCandidatesSharedPtr segments = std::make_shared<SegCompactionCandidates>();
+    if (is_last) {
+        load_noncompacted_segments(segments.get());
+        if (segments->size() == 1) {
+            LOG(INFO) << "only one last candidate segment";
+            rename_compacted_segment_plain(_segcompacted_point);
+            segments->clear();
+        }
+    } else {
+        find_longest_consecutive_small_segment(segments);
+    }
+    return segments;
+}
+
+void BetaRowsetWriter::segcompaction_if_necessary() {
+    if (!config::enable_segcompaction || !config::enable_storage_vectorization) {
+        return;
+    }
+    if (!_is_doing_segcompaction &&
```

Comment:
Two threads can both see `_is_doing_segcompaction` as false, and enter the following block?